### PR TITLE
Making the Tableau in Simplex very lazy

### PR DESCRIPTION
**Pull request by blishko**
This PR introduces quasi-basic variables in the Tableau to keep track of rows that has not been activated yet. This means less work is needed during pivoting as rows which are not active can be ignored. To keep things consistent, care needs to be taken during backtracking to restore a consistent state. Overall, making the Tableau lazy helps quite a bit.

### CODE REVIEW

File: src/tsolvers/lasolver/Simplex.cpp [hunk 1]
```diff
@@ -394,7 +399,8 @@ bool Simplex::checkValueConsistency() const {
     for(unsigned i = 0; i < rows.size(); ++i) {
         if(!rows[i]) {continue;}
         LVRef var {i};
-        if(tableau.isActive(var)){
+        assert(!tableau.isNonBasic(var));
+        if(tableau.isBasic(var)){
```
Comment (aehyvari) [inline src/tsolvers/lasolver/Simplex.cpp]: Maybe add a space between if and (.

File: src/tsolvers/lasolver/Tableau.cpp [hunk 2]
```diff
@@ -18,38 +18,24 @@ namespace {
...
 void Tableau::newNonbasicVar(LVRef v) {
-    assert(!contains(nonbasic_vars, v));
-    while(cols.size() <= v.x) {
-        cols.emplace_back();
-    }
-    while(rows.size() <= v.x) {
-        rows.emplace_back();
-    }
+    assert(!isProcessed(v));
+    ensureTableaReadyFor(v);
```
Comment (aehyvari) [inline src/tsolvers/lasolver/Tableau.cpp]: Is this a typo, should it be ensureTableauReadyFor(v)?

File: src/tsolvers/lasolver/Simplex.cpp [hunk 3]
```diff
@@ -463,6 +468,9 @@ void Simplex::doGaussianElimination( )
 
 Delta Simplex::getValuation(LVRef v) const {
     Delta val(0);
+    if (tableau.isQuasiBasic(v)) {
+        (const_cast<Simplex*>(this))->quasiToBasic(v);
+    }
     if (isRemovedByGaussianElimination(v)) {
```
Comment (aehyvari) [inline src/tsolvers/lasolver/Simplex.cpp]: Is this still a thing, can it happen?

Comment (blishko) [reply src/tsolvers/lasolver/Simplex.cpp]: Gaussian elimination was removed in the next commit. It is not present in the final version, if that is what you were referring to.

File: src/tsolvers/lasolver/Tableau.cpp (hunk 2 above)
Comment (blishko) [reply src/tsolvers/lasolver/Tableau.cpp]: Typo -> fixed

File: src/tsolvers/lasolver/Simplex.cpp (hunk 1 above)
Comment (blishko) [reply src/tsolvers/lasolver/Simplex.cpp]: Fixed

**Comment by aehyvari** [comment]
Right, sorry, my bad!

**Pull request merged**